go/vt/vtgate/buffer/shard_buffer.go

```diff
@@ -569,10 +569,32 @@ func (sb *shardBuffer) drain(q []*entry, err error) {
 	sb.timeoutThread.stop()
 
 	start := sb.timeNow()
-	// TODO(mberlin): Parallelize the drain by pumping the data through a channel.
+
+	// Parallelize the drain by pumping the data through a channel.
+	entryChan := make(chan *entry, len(q))
+
+	parallelism := min(bufferDrainConcurrency, len(q))
+
+	var wg sync.WaitGroup
+	wg.Add(parallelism)
+
+	for i := 0; i < parallelism; i++ {
+		go func() {
+			for _, e := range q {
+				sb.unblockAndWait(e, err, true /* releaseSlot */, true /* blockingWait */)
+			}
+
+			wg.Done()
+		}()
+	}
+
 	for _, e := range q {
-		sb.unblockAndWait(e, err, true /* releaseSlot */, true /* blockingWait */)
+		entryChan <- e
```

Comment:
To elaborate: `entryChan` doesn't need to be buffered because the sending goroutine (i.e. this one) doesn't have anything else to do besides sending the entries through the channel. If it's not blocked on the channel send, it'll block on `wg.Wait()` a couple lines below. So buffering is not an optimization here.